### PR TITLE
test-suite: test GetCSR/SetCertificate

### PR DESCRIPTION
**Pull request by twilfredo**
Add a test to GetCSR, then sign the response and set the generated certificate chain back into slot 1 of the responder.

### CODE REVIEW

File: src/test_suite.rs [hunk 1]
```diff
@@ -271,6 +274,138 @@ pub fn request_all_measurements(cntx: *mut c_void) -> Result<(), u32> {
...
 }
 
+/// # Summary
+///
+/// Request a CSR from the responder, sing it then to set the signed certificate
+/// in `slot_id` of the specified responder established by the `cntx`
+///
+/// # Parameter
+///
+/// * `cntx`: The SPDM context
```
Comment (alistair23) [inline src/test_suite.rs:284]: You are missing `cert_slot_id` here. But if it has to be 0 then why bother passing it?

File: src/test_suite.rs [hunk 2]
```diff
@@ -271,6 +274,138 @@ pub fn request_all_measurements(cntx: *mut c_void) -> Result<(), u32> {
...
+///
+/// # Returns
+///
+/// Ok() Iff there were no error in setting the certificate, `reset-required`
+///      error return case is treated as success.
+/// Panics on any other error in attempt to set-certificate.
+pub fn test_set_certificate(cntx: *mut c_void, cert_slot_id: u8) -> Result<(), ()> {
+    if cert_slot_id >= 8 && cert_slot_id != 0 {
```
Comment (alistair23) [inline src/test_suite.rs:292]: How would it be greater or equal to 8 and equal 0?

File: src/test_suite.rs [hunk 3]
```diff
@@ -271,6 +274,138 @@ pub fn request_all_measurements(cntx: *mut c_void) -> Result<(), u32> {
...
+        .arg("csr_response.cert")
+        .arg("-CA")
+        .arg("./certs/slot0/inter.der")
+        .arg("-sha384")
+        .arg("-days")
+        .arg("3650")
+        .arg("-set_serial")
+        .arg("3")
+        .arg("-extensions")
+        .arg("alias_ca")
```
Comment (alistair23) [inline src/test_suite.rs:351]: -set_serial 2 -extensions device_ca

Comment (alistair23) [reply src/test_suite.rs:351]: Note that this test should fail at the moment

File: src/test_suite.rs [hunk 4]
```diff
@@ -271,6 +274,138 @@ pub fn request_all_measurements(cntx: *mut c_void) -> Result<(), u32> {
...
+        .arg("csr_response.cert")
+        .arg("-out")
+        .arg("csr_response.cert.der")
+        .output()
+        .expect("Failed to execute openssl command")
+        .status
+        .success());
+
+    // 4. Combine all the immutable certificates
+    let immutables_certs = [
```
Comment (alistair23) [inline src/test_suite.rs:372]: Can we run `certs/setup_certs.sh` instead?

File: src/test_suite.rs (hunk 2 above)
Comment (twilfredo) [reply src/test_suite.rs:292]: that should be an || !

**Comment by alistair23** [comment]
This should fail until https://github.com/DMTF/libspdm/pull/2708 is merged

File: src/test_suite.rs (hunk 3 above)
Comment (twilfredo) [reply src/test_suite.rs:351]: This I got from the README, so that should be updated too then?

File: src/test_suite.rs (hunk 4 above)
Comment (twilfredo) [reply src/test_suite.rs:372]: how do you mean? setup_certs.sh doesn't create a bundle with:

```
    let immutables_certs = [
        "./certs/slot0/ca.cert.der",
        "./certs/slot0/inter.cert.der",
        "./csr_response.cert.der",
    ];

```
 into a `set-cert.der` that we can set?

File: src/test_suite.rs (hunk 3 above)
Comment (alistair23) [reply src/test_suite.rs:351]: It should already be fixed

File: src/test_suite.rs (hunk 4 above)
Comment (alistair23) [reply src/test_suite.rs:372]: Ah, you are right. Nevermind

File: src/test_suite.rs (hunk 3 above)
Comment (twilfredo) [reply src/test_suite.rs:351]: `openssl x509 -req -in csr_response.req -out csr_response.cert -CA ./certs/slot0/inter.der -sha384 -days 3650 -set_serial 3 -extensions alias_ca -extfile ./certs/openssl.cnf` Still in the `master` branch under the `Signing a Certificate Signing Request` section.

I have updated that for V2 for this patch.

Comment (alistair23) [reply src/test_suite.rs:351]: Yeah... I just had the change locally. PR pushed

Comment (twilfredo) [reply src/test_suite.rs:351]: ha! nice

**Comment by twilfredo** [comment]
> This should fail until [DMTF/libspdm#2708](https://github.com/DMTF/libspdm/pull/2708) is merged

Added all the changes and yep, failing as of now.

**Comment by alistair23** [comment]
> > This should fail until [DMTF/libspdm#2708](https://github.com/DMTF/libspdm/pull/2708) is merged
> 
> Added all the changes and yep, failing as of now.

Great! Let's leave this open until it passes

**Comment by alistair23** [comment]
Do you mind rebasing this?

**Comment by twilfredo** [comment]
> Do you mind rebasing this?

Done, latest test output.

```
0480: 65 73[2024-06-24T01:53:23Z INFO  spdm_utils::request] Certificate has been set, device requires a restart
[2024-06-24T01:53:23Z INFO  spdm_utils::test_suite] Device Certificate successfully set for slot 1
[2024-06-24T01:53:23Z INFO  spdm_utils::test_suite] Set Certificate ... [OK]
[2024-06-24T01:53:23Z INFO  spdm_utils::test_suite] Testing Complete ...
```

**Comment by twilfredo** [comment]
@alistair23  Unrelated CI failure?

**Review by alistair23 — APPROVED** [review]

**Pull request merged**